### ISSUE TITLE
Enum variants are not being simplified in or-patterns

### ISSUE THREAD

**Issue by nyurik** [rust-lang/rust-clippy#16999]
### Summary

A `Ok(_) | Err(_)` pattern could be simplified to `_` for exhaustive enums.

### Lint Name

unnested_or_patterns

### Reproducer

I tried this code:

```rust
#![warn(clippy::pedantic)]

fn main() {
    let _ = test(Ok(1), Ok(()));
}

fn test(a: Result<u32, i8>, b: Result<(), i8>) -> Result<u32, i8> {
    match (a, b) {
        (Ok(res), Ok(())) => Ok(res),
        (Err(err), Ok(())) | (Ok(_) | Err(_), Err(err)) => Err(err),
    }
}
```

I expected to see this change suggested:

```diff
-        (Err(err), Ok(())) | (Ok(_) | Err(_), Err(err)) => Err(err),
+        (Err(err), Ok(())) | (_, Err(err)) => Err(err),
```

Instead, there were no warnings.

### Version

```text

```

**Comment by tanmay2004** [comment]
Hi, can I work on this?

**Comment by Gri-ffin** [comment]
@tanmay2004 you can use `@rustbot claim` to claim the issue.

**Comment by ronnodas** [comment]
If it affects all exhaustive enums and not just ones from `std` then I would expect this to be a restriction lint, not pedantic.
